File: 231124/우리는 하나/we-are-the-one.cpp
#include <iostream>
#include <algorithm>
#include <queue>

using namespace std;

int n,k,u,d;
int arr[8][8];
int visited[8][8];
queue<pair<int,int>> q;
int result = 0;

int dx[4] = {1,-1,0,0};
int dy[4] = {0,0,1,-1};

// 1. k개 만큼 선택(큐에 넣기)
// 2. 큐가 빌때까지 BFS하며 visited보며 cnt++
// 3. 반환값 중 최댓값 저장

void Print() {
    for(int i = 0; i<n; i++) {
        for(int j = 0; j<n; j++)
            cout << visited[i][j] << ' ';
        cout << endl;
    }
    cout << endl;
}

void VisitClear() {
    for(int i = 0; i<n; i++)
        for(int j = 0; j<n; j++)
            visited[i][j] = 0;
}

bool inRange(int x, int y) {
    return 0 <= x && x < n && 0 <= y && y < n;
}

// step => (arr[x][y] - arr[nx][ny])
bool CanGo(int x, int y, int step) {
    if(!inRange(x,y)) return false;
    if(u > step || step > d) return false;
    if(visited[x][y] == 1) return false;
    return true;
}

void Push(int x, int y) {
    visited[x][y] = 1;
    q.push(make_pair(x,y));
}

void BFS() {
    while(!q.empty()) {
        int x = q.front().first;
        int y = q.front().second;
        q.pop();

        for(int dir = 0; dir < 4; dir++) {
            int nx = x + dx[dir];
            int ny = y + dy[dir];
            if(CanGo(nx,ny, abs(arr[x][y] - arr[nx][ny]))) {
                Push(nx,ny);
            }
        }
    }
}

int VisitCount() {
    int cnt = 0;
    for(int i = 0; i<n; i++)
        for(int j = 0; j<n; j++)
            if(visited[i][j] == 1)
                cnt++;
    return cnt;
}

void Backtracking(int city) {
    if(city == k) {
        BFS();
        // Print();
        // cout << "!" << endl;
        // cout << VisitCount() << endl;
        result = max(result, VisitCount());
        return;
    }

    for(int i = 0; i<n; i++)
        for(int j = 0; j<n; j++) {
            Push(i,j);
            Backtracking(city + 1);
            // Pop(i,j);
            VisitClear();
        }
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n >> k >> u >> d;
    for(int i = 0; i<n; i++)
        for(int j = 0; j<n; j++)
            cin >> arr[i][j];

    Backtracking(0);
    
    cout << result;
    return 0;
}

File: 250221/우리는 하나/we-are-the-one.js
class Queue {
    constructor () {
        this.q = []
        this.head = -1
        this.tail = -1
    }

    empty() {
        return this.head === this.tail
    }

    push(e) {
        this.q.push(e)
        this.tail++
    }

    pop() {
        if (this.empty())
            throw new Error("Queue is Empty")
        return this.q[++this.head]
    }

    size() {
        return this.tail - this.head
    }

    front() {
        if (this.empty())
            throw new Error("Queue is Empty")
        return this.q[this.head + 1]
    }
}

const fs = require('fs')
const input = fs.readFileSync(0).toString().trim().split('\n')

const [n,k,u,d] = input[0].split(' ').map(Number)
const arr = input.slice(1).map(v => v.split(' ').map(Number))

const visited = Array(n).fill().map(v => Array(n).fill(false))
let visited2

const q = new Queue()

const InRange = (x,y) => 0<=x && x<n && 0<=y && y<n;

const CanGo = (x,y) => {
    if(!InRange(x,y)) return false;
    if(visited2[x][y]) return false;
    return true;
}

const BFS = () => {
    const dx = [0,0,1,-1], dy = [1,-1,0,0];

    while(!q.empty()) {
        const [x,y] = q.pop()

        for(let dir = 0; dir < 4; dir++){
            const nx = x + dx[dir], ny = y + dy[dir];

            if(CanGo(nx, ny)) {
                const diff = Math.abs(arr[x][y] - arr[nx][ny])
                if(diff >= u && diff <= d) {
                    q.push([nx,ny])
                    visited2[nx][ny] = true
                }
            }
        }
    }
}

const nodes = []
const Backtracking = () => {
    if(nodes.length === k) {
        visited2 = Array(n).fill().map(v => Array(n).fill(false))
        nodes.forEach(([x,y]) => {
            q.push([x,y])
            visited2[x][y] = true
        })
        BFS()

        let cnt = 0;
        visited2.forEach(arr => arr.forEach(v => v ? cnt++ : ""))

        result = Math.max(result, cnt)
        return
    }

    for(let i = 0; i<n; i++)
        for(let j = 0; j<n; j++) {
            if(visited[i][j]) continue;

            nodes.push([i,j])
            visited[i][j] = true;
            Backtracking()
            visited[i][j] = false;
            nodes.pop()
        }
}

let result = 0
Backtracking()
console.log(result)